File: Google/Google_Interview/3_LinkedList/Medium/Copy_List_with_Random_Pointer/Solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        nodeDict = dict()
        cur = head
        dummyHead = Node(-1)
        copied = dummyHead
        while cur:
            newNode = Node(cur.val)
            copied.next = newNode
            nodeDict[cur] = newNode
            
            copied = copied.next
            cur = cur.next
            
        cur = head
        copied = dummyHead.next
        while cur:
            if cur.random is not None:
                copied.random = nodeDict[cur.random]
                
            copied = copied.next
            cur = cur.next
            
        return dummyHead.next

File: Google/Google_Interview/3_LinkedList/Medium/Copy_List_with_Random_Pointer/Solution.java
import java.util.HashMap;
import java.util.Map;

/*
// Definition for a Node.
*/
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}

class Solution {
    public Node copyRandomList(Node head) {
       
        Map<Node, Node> map = new HashMap<Node, Node>();
        Node curNode = head;
        Node dummyHead = new Node(-1);
        Node copiedHead = dummyHead;
        
        while (curNode != null) {
            Node newNode = new Node(curNode.val);
            copiedHead.next = newNode;
            
            map.put(curNode, newNode);
            
            copiedHead = copiedHead.next;
            curNode = curNode.next;
        }
        
        curNode = head;
        copiedHead = dummyHead.next;
        while (curNode != null) {
            if (curNode.random != null) {
                copiedHead.random = map.get(curNode.random);    
            }
            copiedHead = copiedHead.next;
            curNode = curNode.next;
        }
        
        
        return dummyHead.next;
    }
}